### PR TITLE
Add containsAllInOrder

### PR DESCRIPTION
**Pull request by natebosch**
This matcher is useful for testing that expected values are present
while allowing for additional values, or when used multiple times for
testing the combination of multiple ordered collections while allowing
for aribtrary interleaving.

### CODE REVIEW

File: lib/src/iterable_matchers.dart [hunk 1]
```diff
@@ -263,3 +263,44 @@ class _PairwiseCompare extends _IterableMatcher {
     }
   }
 }
+
+/// Matches [Iterable]s which contain an element matching every value in
+/// [expected] in the same order, but may contain additional values interleaved
+/// throughout.
```
Comment (munificent) [inline lib/src/iterable_matchers.dart]: An example of a matching and non-matching case would be really helpful here.

File: test/iterable_matchers_test.dart [hunk 2]
```diff
@@ -136,6 +136,34 @@ void main() {
...
+        "Which: did not find a value matching a value greater than <0> "
+        "following expected prior values");
+    shouldFail(
+        d,
+        containsAllInOrder([1, 2, 3]),
+        "Expected: contains in order([1, 2, 3]) "
+        "Actual: [0, 1, 0, 2] "
+        "Which: did not find a value matching <3> following expected prior "
+        "values");
+  });
```
Comment (munificent) [inline test/iterable_matchers_test.dart]: Add a test where the object being matched isn't an iterable.

**Comment by natebosch** [comment]
I don't think you can design a sequence that would require backtracking. If multiple values match a given matcher in the expected sequence it's find to consume the early one, then everything up to and including the next one get thrown away as the allowed extra values. You can't have a false negative because of a matcher being too broad.

You can have a false positive if you have a matcher which is too broad and accepts the wrong value in the sequence - but we would expect that an overly broad matcher causes false positives.

File: lib/src/iterable_matchers.dart (hunk 1 above)
Comment (natebosch) [inline lib/src/iterable_matchers.dart]: Done

File: test/iterable_matchers_test.dart (hunk 2 above)
Comment (natebosch) [inline test/iterable_matchers_test.dart]: Done

**Pull request merged**